src/cyclotomic.rs

```diff
@@ -97,80 +99,121 @@ impl Cyclotomic {
+    // change the values represented. Since we mutate, we cannot use the Rust traits
+    // \tt{Eq} and \tt{PartialEq}. The same is true for all arithmetic operations.
+    fn eq(&self, other: &Self) -> bool {
+        let mut z1 = self.clone();
+        let mut z2 = other.clone();
+        Cyclotomic::match_orders(&mut z1, &mut z2);
+        // Since we assume the exponent list is sorted, we can just do a
+        // comparison of the coeff and exp vectors, the number is in
+        // canonical form.
+        z1.coeffs == z2.coeffs && z1.exps == z2.exps
+    }
 
-        while z2_i < z2.exps.len() && z2.exps[z2_i] < z1.exps[z1_i] {
-            result.coeffs.push(z2.coeffs[z2_i].clone());
-            result.exps.push(z2.exps[z2_i].clone());
-            z2_i += 1;
-        }
+    fn ne(&self, other: &Self) -> bool {
+        return !eq(self, other);
+    }
+}
 
-        // If the exponents match up, we just add the coefficients - they are for the same term
-        if z1.exps[z1_i] == z2.exps[z2_i] {
-            result
-                .coeffs
-                .push(z1.coeffs[z1_i].clone() + z2.coeffs[z2_i].clone());
-            result.exps.push(z1.exps[z1_i]);
-            z1_i += 1;
-            z2_i += 1
+impl Add for Cyclotomic {
+    type Output = Cyclotomic;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let mut z1 = self.clone();
+        let mut z2 = rhs.clone();
+        Cyclotomic::match_orders(&mut z1, &mut z2);
+
+        let mut z1_i = 0;
+        let mut z2_i = 0;
+        let mut result = Cyclotomic::zero();
+
+        // We are producing a result cyclotomic in canonical form assuming the
+        // inputs are in canonical form. That is, we build the result term by
+        // term, assuming that the exponent lists are sorted.
+        while z1_i < z1.exps.len() || z2_i < z2.exps.len() {
+            while z1_i < z1.exps.len() && z1.exps[z1_i] < z2.exps[z2_i] {
+                result.coeffs.push(z1.coeffs[z1_i].clone());
+                result.exps.push(z1.exps[z1_i].clone());
+                z1_i += 1;
+            }
+
+            while z2_i < z2.exps.len() && z2.exps[z2_i] < z1.exps[z1_i] {
+                result.coeffs.push(z2.coeffs[z2_i].clone());
+                result.exps.push(z2.exps[z2_i].clone());
+                z2_i += 1;
+            }
+
+            // If the exponents match up, we just add the coefficients - they are for the same term
+            if z1.exps[z1_i] == z2.exps[z2_i] {
+                result
+                    .coeffs
+                    .push(z1.coeffs[z1_i].clone() + z2.coeffs[z2_i].clone());
+                result.exps.push(z1.exps[z1_i]);
+                z1_i += 1;
+                z2_i += 1
+            }
         }
+
+        result
     }
+}
 
-    result
+impl Mul for Cyclotomic {
+    type Output = Cyclotomic;
+
+    // I could have a guess at how to implement this, but I would inevitably embarrass myself.
```

Comment:
There is a paper by some guy with some basic algorithms for this: https://dl.acm.org/doi/pdf/10.1145/1086837.1086847. The thing is he's talking about polynomial rings, not necessarily quotiented by anything. So there are a lot of reductions we can do.